### PR TITLE
Record: Scylla + GPTQ + BH3072 — val_bpb 1.0856 (3-seed mean)

### PR DESCRIPTION
**Pull request by anthony-maio**
## Summary

- **val_bpb: 1.0856** (3-seed mean)
- **Artifact:** 15.3-15.8 MB (all seeds < 16MB)
- **Training:** 600s on 8xH100 SXM | No SLOT, No TTT

## 3-Seed Results

| Seed | Sliding BPB | Artifact |
|------|------------|----------|
| 1337 | 1.1009 | 15,267,156 |
| 42 | **1.0782** | 15,813,568 |
| 2024 | **1.0777** | 15,807,116 |
| **Mean** | **1.0856** | |

Beats merged SOTA (1.1147, PR #1019) by **0.029 BPB** (14x significance threshold).

## Key Techniques

- **Scylla tokenizer** (998-vocab TokenMonster, PR #1143): 37% fewer tokens/byte
- **AR self-gen Full Hessian GPTQ** (int6): Cholesky error compensation, 64 self-gen seqs
- **BigramHash 3072x112**, **VRL**, **XSA all 11 layers**, **QK-Gain 4.0**
- EMA(0.997) + SWA, Late QAT, LZMA-9, FA3

## Compliance

- **No SLOT** — no eval-time delta optimization
- **No TTT** — no eval-time weight updates
- No n-gram cache, no network calls
- Tokenizer byte accounting via validated metadata (candidate.meta.npz)
- All artifacts under 16MB, all training under 600s

## Note on seed variance

Seed 1337 trained at ~153ms/step (pod throttling), getting only 3933 steps vs 6500+ for seeds 42/2024. Seeds 42/2024 are the representative results at 1.078 BPB. A clean 3-seed on a consistently fast pod would yield mean ~1.079.

## Reproduction

```bash
VOCAB_SIZE=998 BIGRAM_VOCAB_SIZE=3072 BIGRAM_DIM=112 WARMDOWN_ITERS=4000 \
DATA_PATH=./data/datasets/fineweb10B_scylla \
TOKENIZER_PATH=./candidate.vocab TOKENIZER_META_PATH=./candidate.meta.npz \
SEED=1337 torchrun --standalone --nproc_per_node=8 train_gpt.py
```

Scylla shards: `anthonym21/fineweb10B-scylla` on HuggingFace.

## Credits

- Scylla tokenizer: @simon-marcus (PR #1143)
- Training stack: PR #175, PR #1019 (@abaybektursun)
- GPTQ: PR #1019 (@abaybektursun)
- VRL: ResFormer (arXiv:2410.17897)

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Record: Scylla + GPTQ + BH3072 — val_bpb 1.0856 (3-seed mean)

**Compliance:** LOOKS CLEAN — legal score-first-per-chunk TTT (PR #1413 pattern)

## PR #1405 — Scylla_GPTQ_BH3072 (Anthony Maio / anthony-maio) ### N-gram / BigramHash Family Bug Check BigramHash is implemented at lines 456–477. The hash at line 471 is: `out[..., 1:] = torch.bitwise_xor(36313 * t[..., 1:], 27191 * t[..., :-1]) % mod` where `t = tokens.to(torch.int32)` — i.e., the *input* tokens (x), not the targets (y). The target is never XORed into the hash key. **No ILLEGAL n-gram family bug present.** ### Pre-Quant TTT Check No Pre-Quant TTT pattern detected. Training is completed before quantization; no multi-epoch training loop runs on `val_tokens` during or before scoring in the post-train phase. Training ends, GPTQ int6 quantization is applied, then evaluation and optional TTT proceed. **No ILLEGAL Pre-Quant TTT.** ### Score-First TTT Check `eval_val_sliding_ttt` (lines 766–880) implements legal score-first TTT: - Phase 1 (lines 809–835): `base_model.eval()` + `torch.inference_mode()` — every window in the chunk is scored with the current model state and loss accumulated BEFORE any weight update. - Phase 2 (lines 836–865): `if ci != num_chunks - 1` guard ensures the **last chunk is never trained on** — exactly the `is_last_chunk` guard. Training with `ttt_epochs` loops only applies to non-final chunks using `val_tokens[chunk_start:chunk_end]`. - Score is accumulated before any gradient step in every chunk. This matches the PR #1413 legal score-first pattern. **LEGAL score-first TTT confirmed.** ### OGD Check `run_ogd_sliding` (lines 881–940) scores batches first with `torch.inference_mode()`, accumulates NLL, then updates `ogd_bias` token-by-token *within* the same batch loop (lines 929–936). OGD bias update happens after the NLL for that token position has...

**Verdict:** LOOKS CLEAN — legal TTT implementation matching the PR #1413 (dexhunter) pattern: each chunk scored under `torch.no_grad()` before `optimizer.step()`, with `is_last_chunk` guard preventing adaptation on the final scored chunk.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). TTT implementation follows the legal score-first discipline.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
